New `spack.environment.active_environment` api, and make spack.environment not depend on spack.cmd.

haampie opened the pull request:
- Rework `spack.environment.get_env(...)` to `spack.environment.active_environment()`
  which is now a trivial getter for the active environment.
- Introduce `spack.cmd.require_active_env(cmd_name=...)` which is a wrapper 
  around `spack.environment.active_environment()` but errors with a helpful
  error message when used from the CLI.
- Clean up instances where former `spack.environment.get_env` was abused
outside the context of a command (fake command name, fake `-e` flag, 
  bunch of `None`, `None` arguments)
- Remove the `-e` parsing from former `get_env`, because `main.py` is
responsible for processing `-e` and already activates the environment.
  In fact this logic was entirely broken, because it did not respect `-D` at all
  for loading an env from a directory, and it did not respect `-E` (i.e. it would
  load an environment even when explicitly disabled).
- Move `spack.environment.find_environment` to
`spack.cmd.find_environment`, to avoid that spack.environment is aware
of argparse.
- Move command line argument parsing of environment `installs_specs` from 
  `spack.environment` to `spack.cmd`, to avoid that spack.environment is aware
  of argparse.
- Remove some unused imports, and work around circular imports.
- Use `ev` alias for `spack.environment` consistently where it works Python 2.7
  permitting

TL;DR: better interface for getting the currently active environment, and refactor
things so that `spack.environment` does not depend on `spack.cmd`.

haampie committed:
Refactor active environment getter

- Make `spack.environment.get_env` a trivial getter for the active
environment.
- Introduce `spack.cmd.get_env_for_command(cmd_name, env)` which is a
wrapper around spack.env.get_env(env)` but adds support for an error
message for the particular command.
- Clean up instances where `spack.environment.get_env` was abused
outside the context of a command (fake command name, bunch of None
arguments which are hard to follow)
- Remove the `-e` parsing from `get_env`, because `main.py` is
responsible for processing `-e` and already activates the environment.
- Move `spack.environment.find_environment` to
`spack.cmd.find_environment`, to avoid that spack.environment is aware
of argparse.

haampie committed:
Fix reference to Environment for docs

haampie committed:
get_env -> get_active_env, add cmd.require_env

`spack.environment.get_env` -> `spack.environment.get_active_env`
`spack.cmd.get_env_for_command` -> `spack.cmd.require_env`

Use `get_active_env` when no environment is required, and make
`require_env` always require an active environment.

haampie commented:
Added the `breaking-change` label for posterity because `spack.environment.get_env` got removed.

haampie commented:
@gartung any idea why `test_default_rpaths_create_install_default_layout` is failing? restarted the tests a couple times but it keeps failing.

gartung commented:
Stage directory was not removed
https://github.com/spack/spack/blob/37731856395699956093d02a6ff64bf7cbf0f969/lib/spack/spack/test/conftest.py#L291

haampie commented:
Sure, but it's failing spuriously, any idea why?

gartung commented:
It's a test package so it creates it's own files on the fly. Maybe there is a timing issue with the cleanup of the stage directory.

becker33 commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -501,3 +501,71 @@ def extant_file(f):
     return f
+
+
+def require_env(cmd_name):
```
This function should be `require_active_env`, because `require_env('add')` sounds like it's requiring an environment named "add", not requiring that an environment be active and passing a command name.

When we call this method, we should use the kwarg calling convention for clarity. `require_active_env(cmd_name='add')` is a lot clearer in this case.

becker33 commented on lib/spack/spack/cmd/add.py:
```diff
@@ -22,7 +21,7 @@ def setup_parser(subparser):
 
 def add(parser, args):
-    env = ev.get_env(args, 'add', required=True)
+    env = spack.cmd.require_env('add')
```
See above, and same for all the similar instances.

becker33 commented on lib/spack/spack/cmd/verify.py:
```diff
@@ -74,7 +73,7 @@ def verify(parser, args):
     elif args.specs_or_files:
         # construct disambiguated spec list
-        env = ev.get_env(args, 'verify')
+        env = spack.environment.get_active_env()
```
You need to import `spack.environment` at the top here, and this goes for most of the files where you deleted it.

The only reason it's working without is because the name gets bound from importing `spack.store`, but you shouldn't rely on that because it's undefined python behavior.

becker33 replied:
@haampie I think this might be related to the test failure -- you can get some really inscrutable errors with improper imports.

haampie replied:
Hm okay, let me try to fix the circular dependencies. Note that `def matching_spec_from_env` in `__init__.py` already used `spack.environment.<stuff>` without importing before this pr.

But I mean, this PR is intended to separate core domain stuff (spack.environment) from the interface layer (spack.cmd), so optimally there would not be any circular deps at all...

haampie replied:
Couple of commits were necessary to fix circular imports, but then it's OK to add those `spack.environment` imports in spack.cmd.*. I've done it over the whole code base in 1ac0eb33d29ef06e45f6ccc102232a3b45adab0c, and made it consistently use `import spack.environment as ev` since this was already done in 55 out of 62 cases.

becker33 reviewed (changes requested):
I don't love the arrangement of methods in the cmd vs env modules, but I didn't love the old arrangement either so I'm not going to complain too much about that.

A couple minor issues.

haampie committed:
Move config/compiler imports into function

Undo global imports from https://github.com/spack/spack/pull/21442,
they seem to introduce circular dependencies.

haampie committed:
Add missing imports, and use ev as shorthand everywhere for consistency

haampie committed:
Remove last dep of spack.environment on spack.cmd

The way I see it update_kwargs_from_args could have been a constructor
for a PackageInstallArgs object, and they could have a named constructor
like PackageInstallArgs::from_cli_args(args). So it makes sense to move
this function to spack.installer, and with that the last dependency of
spack.environment on spack.cmd is gone, as it should be.

haampie committed:
spack.cmd.require_env -> spack.cmd.require_active_env

haampie committed:
spack.cmd.elide_list -> spack.util.elide_list

haampie committed:
Remove unused spack.cmd import

becker33 commented on lib/spack/spack/util/__init__.py:
```diff
@@ -2,3 +2,19 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+def elide_list(line_list, max_num=10):
```
This probably belongs in llnl.util.lang (the same module as `dedupe` and `star`)

haampie replied:
done

becker33 commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -501,3 +486,71 @@ def extant_file(f):
     return f
+
+
+def require_active_env(cmd_name):
```
I don't fully understand why this method would be in spack.cmd... does that help with the circular import problem if you put it in `spack.environment` (and leave `find_environment` here because it uses argparse)?

haampie replied:
It's in `spack.cmd` because its only purpose is to create an error that shows you how to use the command if you didn't activate an env.

haampie commented:
@spackbot run pipeline

alalazo commented on lib/spack/spack/environment.py:
```diff
@@ -263,100 +264,9 @@ def deactivate(shell='sh'):
-            '    spack env activate ENV',
-            'or use:',
-            '    spack -e ENV %s ...' % cmd_name)
+def get_active_env():
+    """Returns the active environment when there is any"""
+    return _active_environment
```
Since this is a refactoring why not avoiding the `get_` prefix and call the function `active_environment`?

haampie replied:
Sounds good to me. Not sure what standard practices are in python-land, but we could as well drop the getter and rename _active_environment to active_environment and use it directly?

alalazo replied:
I thought about that and it's certainly an option. If we use a function call we can later extend it with additional arguments with defaults without breaking the previous API.

alalazo commented on lib/spack/spack/installer.py:
```diff
@@ -613,6 +613,42 @@ def package_id(pkg):
     return "{0}-{1}-{2}".format(pkg.name, pkg.version, pkg.spec.dag_hash())
 
 
+def update_kwargs_from_args(args, kwargs):
```
I feel like this is the wrong place for this function if the aim was to have the spack core module not dependent on the cli. Here we assume knowledge on which arguments are passed in an `argparse` namespace.

haampie replied:
Yeah, I tried to explain this in the commit message 89177317cc2b50e2845bf968507c8afa3a0ab0a5:

> The way I see it update_kwargs_from_args could have been a constructor
> for a PackageInstallArgs object, and they could have a named constructor
> like PackageInstallArgs::from_cli_args(args). So it makes sense to move
> this function to spack.installer, and with that the last dependency of
> spack.environment on spack.cmd is gone, as it should be.

haampie replied:
I could move it out of installer.py, but I don't know where it should go instead :thinking: installer_args.py?

alalazo replied:
Ok, I see that commit. My opinion is that the right solution is to have:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/environment.py#L1575

and similar functions not having the `args` argument. It should be on the caller of the interface to prepare the `install_args` appropriately.

alalazo replied:
Looking at:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/cmd/install.py#L197

it seems this is doable. Besides, `update_kwargs_from_args` is called _internally_ from `Environment.install_specs` which receives directly the `cli_args` and externally in the function above. So why not setting everything in `cmd/install.py` and just pass a dictionary after a single call to the function?

haampie replied:
I've removed `args` from `spack.environment.install_spec(...)` and `spack.environment.install_all(...)` and combined the `update_kwargs_from_args` in `spack.cmd.install`. That seems like a better solution indeed.

alalazo replied:
That's exactly what I had in mind. Thanks!

alalazo commented on lib/spack/spack/test/cmd/bootstrap.py:
```diff
@@ -7,7 +7,7 @@
 
 import spack.config
-import spack.environment
+import spack.environment as ev
```
Why this change?

haampie replied:
See https://github.com/spack/spack/pull/25439#issuecomment-900906456

alalazo replied:
I think we can take care of understanding these two failures in a following PR.

alalazo commented on lib/spack/spack/architecture.py:
```diff
@@ -587,6 +585,8 @@ def __call__(self):
 
 @contextlib.contextmanager
 def use_platform(new_platform):
+    import spack.compilers
+    import spack.config
```
Does this mean that this refactor introduces new circular dependencies at module creation? Is this due to some function that moved from one module to another?

haampie replied:
Hm, weird, by now it's no longer necessary... Maybe because I also deleted some imports elsewhere afterwards.

alalazo commented on lib/spack/spack/concretize.py:
```diff
@@ -63,10 +63,11 @@ def __init__(self, abstract_spec=None):
         self._adjust_target_answer_generator = None
 
     def concretize_develop(self, spec):
+        import spack.environment as ev
```
Why doing a local import rather than one at the top of the module?

haampie replied:
Because of Python 2:
```
$ SPACK_PYTHON=python2 spack --version
Traceback (most recent call last):
  File "/home/harmen/spack/bin/spack", line 96, in <module>
    import spack.main  # noqa
  File "/home/harmen/spack/lib/spack/spack/main.py", line 36, in <module>
    import spack.cmd
  File "/home/harmen/spack/lib/spack/spack/cmd/__init__.py", line 24, in <module>
    import spack.environment as ev
  File "/home/harmen/spack/lib/spack/spack/environment.py", line 20, in <module>
    import spack.concretize
  File "/home/harmen/spack/lib/spack/spack/concretize.py", line 36, in <module>
    import spack.environment as ev
AttributeError: 'module' object has no attribute 'environment'
```

There's a circularity between environment & concretize, but it's only an error on Python 2.7, not Python 3.x.

haampie replied:
Actually... when I drop `as ev` here, Python 2 does not complain :confused:.

alalazo replied:
Can you open an issue as soon as we merge this, to understand what's going on?

alalazo commented on lib/spack/spack/package.py:
```diff
@@ -1519,6 +1519,7 @@ def all_patches(cls):
         return patches
 
     def content_hash(self, content=None):
+        import spack.environment as ev
```
Why a local import?

alalazo reviewed (commented):
I left a few comments. Overall I think this is an improvement over current state, and it catches a few places where we were missing imports. There's a function that I think is being moved to the wrong module and a few local imports that I would avoid if possible.

Style question to everyone. Do we want to consistently:
```python
import spack.environment as ev
```
everywhere? I have a slight preference for fully qualified names, but if we go with an alias everywhere I'm fine with it. It would be good if we can decide on that and have the decision checked automatically by style tests, if possible and easy to do.

haampie commented:
> Do we want to consistently:
> 
> ```python
> import spack.environment as ev
> ```
> 
> everywhere?

My reason for it was that we're doing it already in 88% of the cases:

```console
$ git checkout develop
$ grep '--include=*.py' 'import spack.environment as ev' -r lib/ | wc -l
55
$ grep '--include=*.py' 'import spack.environment' -r lib/ | wc -l
62
```

Edit: turns out python 2.7 struggles with alias'ed imports 6b8e412dcec00acd314ec48e942140121ffaf0b0, so `as ev` is used everywhere except in two cases now.

haampie committed:
Make more imports top-level (review comments)

Apparently no longer necessary to have local imports because of later
changes

haampie committed:
Python 2: toplevel import errors only with 'as ev'

The error message "'module' object has no attribute 'environment'" is
gone after using `import spack.environment` without an alias on Python
2.7, so these are the only exceptions to `as ev` everywhere...

haampie committed:
spack.environment.get_active_env() -> spack.environment.active_environment()

haampie committed:
Revert "Remove last dep of spack.environment on spack.cmd"

This reverts commit 89177317cc2b50e2845bf968507c8afa3a0ab0a5.

haampie committed:
Handle cli args -> package installer args in cmd, not in environment

haampie commented:
I think I've incorporated all the review comments now, and I've updated the PR description at the top with a summary of the changes.

alalazo commented on lib/spack/spack/cmd/install.py:
```diff
@@ -324,10 +324,14 @@ def get_tests(specs):
 
+    # Parse cli arguments and construct a dictionary
+    # that will be passed to the package installer
+    update_kwargs_from_args(args, kwargs)
```
This looks much better to me now!

alalazo commented on lib/spack/spack/environment.py:
```diff
@@ -1560,21 +1470,18 @@ def uninstalled_specs(self):
-        from spack.installer import PackageInstaller
+        self.install_specs(None, **install_args)
 
+    def install_specs(self, specs=None, **install_args):
```
Definitely a better API imo without a dependency on an `argparse` namespace. Thanks for taking care of that!

alalazo reviewed (dismissed):
LGTM but let's wait for @becker33 review. Should we wait on #22845 and #22720 to be merged before merging this PR, given  the number of files it touches?

becker33 reviewed (changes requested):
Mostly looks good, I agree the changes you made after Max's review are great improvements. only minor stuff left.

haampie committed:
spack.util.elide_list -> llnl.util.lang.elide_list

becker33 reviewed (approved):
I'm still not convinced on whether the methods are all in the perfect modules, but I don't think we should hold this up over it.

haampie commented:
Thanks!
